contrib/src/main/java/com/datatorrent/contrib/cassandra/AbstractUpsertOutputOperator.java

```diff
@@ -0,0 +1,1401 @@
+
+  transient Map<DataType.Name, List<String>> dataTypeGroupedColumns;
+
+  transient Set<String> pkColumnNames;
+
+  transient Set<String> counterColumns;
+
+  transient Set<String> collectionColumns;
+
+  transient Set<String> listColumns;
+
+  transient Set<String> mapColumns;
+
+  transient Set<String> setColumns;
+
+  transient Set<String> userDefinedTypeColumns;
+
+  transient Set<String> regularColumns;
+
+  transient Map<String, String> pojoFieldNamesToCassandraColumns;
+
+  Map<String, Object> getters;
+
+  Map<String, TypeCodec> codecsForCassandraColumnNames;
+
+  transient Map<Long, PreparedStatement> preparedStatementTypes;
+  transient Class<?> tuplePayloadClass;
+
+  private transient Logger LOG = LoggerFactory.getLogger(AbstractUpsertOutputOperator.class);
+
+  // Metrics
+
+  @AutoMetric
+  transient long ignoredRequestsDuetoIfExistsCheck = 0;
+
+  @AutoMetric
+  transient long successfullWrites = 0;
+
+  @AutoMetric
+  long totalWriteRequestsAttempted = 0;
+
+  @AutoMetric
+  transient int numberOfHostsWrittenTo = 0;
+
+  @AutoMetric
+  transient double fifteenMinuteWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient double fiveMinuteWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient double oneMinuteWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient double meanWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient long totalIgnoresInThisWindow = 0;
+
+  @AutoMetric
+  long totalIgnoresSinceStart = 0;
+
+  @AutoMetric
+  transient long totalWriteTimeoutsInThisWindow = 0;
+
+  @AutoMetric
+  long totalWriteTimeoutsSinceStart = 0;
+
+  @AutoMetric
+  transient long totalWriteRetriesInThisWindow =  0;
+
+  @AutoMetric
+  long totalWriteRetriesSinceStart = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyOne = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyTwo = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyThree = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyAll = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyLocalOne = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyQuorum = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyLocalQuorum = 0;
+
+  @AutoMetric
+  transient long writeWithConsistencyLocalSerial = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyEachQuorum = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencySerial = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyAny = 0;
+
+  transient Set<Host> uniqueHostsWrittenToInCurrentWindow;
+
+
+  @InputPortFieldAnnotation
+  public final transient DefaultInputPort<UpsertExecutionContext> input = new DefaultInputPort<UpsertExecutionContext>()
+  {
+    @Override
+    public void process(final UpsertExecutionContext tuple)
+    {
+      if (!isEligbleForPassivation(tuple)) {
+        return;
+      }
+      BoundStatement stmnt = setDefaultsAndPrepareBoundStatement(tuple);
+      ResultSet result = session.execute(stmnt);
```

Comment:
Are we updating one row at a time? I understand using batches and also provide a guarantee of exactly once processing is not easy. But hitting db for each query can also be expensive. Any thoughts on this?